README.md
# Memoriae

Anonymously post your thoughts and memoriaes

## Installation & Usage

### Installation

- Clone or download the repo

### Usage

- Go to the project folder
- Open the bash terminal
- Run: `bash _scripts/startDev.sh`
- You will find:
  - Client in [localhost:3000](http://localhost:3000/) 
  - Server in [localhost:5000](http://localhost:5000/)
- To close the app:
  - Open a new bash terminal in the project folder
  - Run: `bash _scripts/stop.sh`
- To perfom a complete teardown of the app:
  - Open a new bash terminal in the project folder
  - Run: `bash _scripts/teardown.sh`

## Technologies

<details>
  <summary><b>Environment</b></summary>

- [Docker](https://www.docker.com/)

</details>

<details>
  <summary><b>Client</b></summary>
  
  - HTML
  - CSS
  - JavaScript
  - [NPM](https://www.npmjs.com/)
    - [lite-server](https://www.npmjs.com/package/lite-server)
    - [concurrently](https://www.npmjs.com/package/concurrently)
    - [watchify](https://www.npmjs.com/package/watchify)

</details>

<details>
  <summary><b>Server</b></summary>

- [NodeJs](https://nodejs.org/en/)
- [NPM](https://www.npmjs.com/)
  - [express](https://www.npmjs.com/package/express)
  - [cors](https://www.npmjs.com/package/cors)
  - [cross-env](https://www.npmjs.com/package/cross-env)
  - [morgan](https://www.npmjs.com/package/morgan)
  - [uuid](https://www.npmjs.com/package/uuid)
  - [nodemon](https://www.npmjs.com/package/nodemon)
  - [pg](https://www.npmjs.com/package/pg)

</details>

<details>
  <summary><b>Database</b></summary>

- [PostgreSQL](https://www.postgresql.org/)

</details>

## API

| Route Name | URL        | HTTP Verb | Description             |
| ---------- | ---------- | --------- | ----------------------- |
| Root       | /          | GET       | Test api is initialised |
| Create     | /posts     | POST      | Create a post           |
| Show       | /posts/:id | GET       | Show info about a post  |

## Database Schema

![Database Schema](./readme_assets/db_schema.jpg)

## Changelog

Find the full changelog [here](./changelog.md)

## Bugs

- [ ]  Change the system to make friendly URL
- [ ]  When you search some address in the url and you add extra slash parameters to the home address the 404 page doesn't show.
- [ ]  The default 'Anonymous' of the user is not working with the actual insert query when author is empty or null (for now it's been asign before the query).

## Licence

[MIT Licence](https://opensource.org/licenses/mit-license.php)


changelog.md
# Changelog

## Client

- stub file structure
- add initial dependencies
  - lite-server
  - concurrently
  - watchify
- add favicon
- render new post form at `/`
- render post at `/#id`
- impliment new post sumbission with form
- fix field casing in createFormHTML helpers function
- add frontend styling
- add 404 page for unfound posts
- make frontend responsive

## Server

- stub file structure
- add initial dependencies
  - express
  - morgan
  - cors
  - nodemon
  - uuid
  - pg
- create basic express server
- added init.js into dbconfig
- create db example case
- added the db seed part to package.json
- Model created
- app router function created
- fix post constructor and create sql query
- add id validtation to `GET /posts/:id`
- add default author of "Anonymous" to post model create method
- adapted the access to the db for Heroku
- added NODE_ENV configuration

## Database

- Stub setup file
- Filled setup file

## Misc

- Add Readme Sections and content where possible
- Populate docker-compose file
- Docker composer now execute 'npm run seedDev'
- Docker compose now includes the client
- Added the Start/Teardown scripts and include them in the readme
- Added Stop script and his part on the readme

## Stretch Goals

- Deploy to Heroku
- Testing
- Edit posts
- Delete posts
- User friendly route names on client


client/static/bundle.js
(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
const { initPageBindings } = require('./lib/handlers');

initPageBindings();

},{"./lib/handlers":3}],2:[function(require,module,exports){
const serverUrl = require('./url');

async function getPostData(id) {
	try {
		const url = `${serverUrl}posts/${id}`;
		const response = await fetch(url);
		const postData = await response.json();
		return postData;
	} catch (err) {
		console.error(err);
	}
}

async function createNewPost(body) {
	try {
		const headers = new Headers();
		headers.append('Content-Type', 'application/json');
		const options = {
			method: 'POST',
			headers,
			body: JSON.stringify(body),
		};
		const url = `${serverUrl}posts/`;
		const response = await fetch(url, options);
		const newPost = response.json();
		return newPost;
	} catch (err) {
		console.error(err);
	}
}

module.exports = { getPostData, createNewPost };

},{"./url":5}],3:[function(require,module,exports){
const main = document.querySelector('main');
const { getPostData, createNewPost } = require('./fetch');
const { createFormHTML, createPost, renderPostNotFound } = require('./helpers');

async function handlePostFormSubmit(e) {
	try {
		e.preventDefault();
		const body = Object.fromEntries(new FormData(e.target));
		const postData = await createNewPost(body);
		if (!postData.id) {
			throw new Error('New post has no id.');
		}
		window.location.hash = postData.id;
	} catch (err) {
		console.error(err);
	}
}

async function updateMain(hash) {
	main.textContent = '';
	let child;
	if (hash) {
		const postData = await getPostData(hash.slice(1));
		if (!postData.id) {
			renderPostNotFound();
			return;
		}
		child = createPost(postData);
	} else {
		child = createFormHTML(handlePostFormSubmit);
	}
	main.append(child);
}

function updateConent() {
	const hash = window.location.hash;
	updateMain(hash);
}

function initPageBindings() {
	document.addEventListener('DOMContentLoaded', updateConent);
	window.addEventListener('hashchange', updateConent);
}

module.exports = { initPageBindings };

},{"./fetch":2,"./helpers":4}],4:[function(require,module,exports){
function createFormHTML(eventHandler) {
	const form = document.createElement('form');
	form.autocomplete = 'off';
	const fields = ['title', 'author', 'body'];

	fields.forEach((field) => {
		const inputLabel = document.createElement('label');
		inputLabel.for = field;
		inputLabel.innerText = field;
		form.append(inputLabel);

		const input = document.createElement('input');
		input.type = 'text';
		input.id = field;
		input.name = field;
		input.placeholder = field;

		if (field === 'title') {
			input.required = true;
			input.maxLength = 100;
		} else if (field === 'author') {
			input.maxLength = 100;
		} else {
			input.required = true;
			input.maxLength = 1000;
		}
		form.append(input);
	});

	const submit = document.createElement('input');
	submit.type = 'submit';
	submit.value = 'Add Memoriae';
	form.append(submit);

	form.addEventListener('submit', eventHandler);

	return form;
}

function createPost({ title, author, body, date }) {
	const post = document.createElement('article');

	const header = document.createElement('h2');
	header.innerText = title;
	post.append(header);

	const postAuthor = document.createElement('p');
	postAuthor.innerText = author;
	post.append(postAuthor);

	const postDate = document.createElement('p');
	postDate.innerText = new Date(date).toDateString();
	post.append(postDate);

	const postBody = document.createElement('p');
	postBody.innerText = body;
	post.append(postBody);

	return post;
}

function renderPostNotFound() {
	const main = document.querySelector('main');

	const header = document.createElement('h2');
	header.innerText = '404';

	const message = document.createElement('p');
	message.innerText = 'Memoriae does not exist';

	const article = document.createElement('article');

	article.append(header);
	article.append(message);
	main.append(article);
}
module.exports = { createFormHTML, createPost, renderPostNotFound };

},{}],5:[function(require,module,exports){
const url =
	window.location.hostname === 'localhost'
		? 'http://localhost:5000/'
		: 'https://memoriaes.herokuapp.com/';

module.exports = url;

},{}]},{},[1]);
